### DISCUSSION TITLE
Applying Treesitter to arbitary text as a component of a nvim plugin

### DISCUSSION THREAD

**Discussion by rktjmp** [nvim-treesitter/nvim-treesitter#955] [Q&A] [answered]
There is a plugin called ["Emmet"](https://docs.emmet.io/cheat-sheet/), that expands some text into HTML. I would like to make something similar for another language.

Before I start trying to write an LPEG grammar (assuming that can do what I want...), I thought maybe I could write the grammar as a treesitter module (which seems much nicer to use so far, mostly due to its natural test framework), then pass that parsed AST to my lua plugin to perform the code expansion.

I would only want to run the ts-grammar on a single line/selection/range since the rest of the file would be in a different language. I'm not sure I can see a way to do that in the docs.

Basically something like:

```
text = get_user_text()
tree = treesitter.load_grammar("my-content-type").parse(text)
for node in tree
  generate text for node type
  recurse(node)
```

Is this possible to do?

**Comment by theHamsta** [comment, marked as answer]
Hi. This is totally possible just have a look at the tree-sitter documentation of Neovim or the tree-sitter tests in the Neovim repo. You can call parser:parser() either with a buffer number or with a string (on mobile I can't look up the exact spot in the documentation).

You can also use the html parser within a document of another language so you maybe don't have to parse string. There are parsers/extensions and PR to support angular/vue/svelte

**Reply by rktjmp** [reply]
Ah, whoops. I was looking at this projects docs, didn't think any TS stuff was in nvim-actual yet.

For anyone else stumbling through: https://neovim.io/doc/user/treesitter.html

**Reply by theHamsta** [reply]
I guess the docs are not up to date yet. You can in almost all positions of `bufnr` also use a string.
`parser = vim.treesitter.get_parser(bufnr, lang)`
